File: vap_player_android/android/src/main/kotlin/app/misi/vap_player_android/QueuingEventSink.kt
package app.misi.vap_player_android

/**
 * An event sink that queues events until a downstream delegate attaches,
 * so events emitted before Dart subscribes are not lost.
 */
class QueuingEventSink<T> {
    private var delegate: PigeonEventSink<T>? = null
    private val eventQueue = ArrayList<Any>()
    private var done = false

    private class EndOfStreamEvent

    fun setDelegate(delegate: PigeonEventSink<T>?) {
        this.delegate = delegate
        maybeFlush()
    }

    fun endOfStream() {
        enqueue(EndOfStreamEvent())
        maybeFlush()
        done = true
    }

    fun success(event: T) {
        enqueue(event as Any)
        maybeFlush()
    }

    private fun enqueue(event: Any) {
        if (done) return
        eventQueue.add(event)
    }

    private fun maybeFlush() {
        val delegate = this.delegate ?: return
        for (event in eventQueue) {
            if (event is EndOfStreamEvent) {
                delegate.endOfStream()
            } else {
                @Suppress("UNCHECKED_CAST")
                delegate.success(event as T)
            }
        }
        eventQueue.clear()
    }
}

File: vap_player_ios/ios/Classes/QueuingEventSink.swift
import Foundation

/// An event sink that queues events until a downstream delegate attaches,
/// so events emitted before Dart subscribes are not lost.
final class QueuingEventSink<T> {
  private var delegate: PigeonEventSink<T>?
  private var queue: [Any] = []
  private var done = false

  private struct EndOfStream {}

  func setDelegate(_ delegate: PigeonEventSink<T>?) {
    self.delegate = delegate
    maybeFlush()
  }

  func endOfStream() {
    enqueue(EndOfStream())
    maybeFlush()
    done = true
  }

  func success(_ event: T) {
    enqueue(event)
    maybeFlush()
  }

  private func enqueue(_ event: Any) {
    if done { return }
    queue.append(event)
  }

  private func maybeFlush() {
    guard let delegate = delegate else { return }
    for event in queue {
      if event is EndOfStream {
        delegate.endOfStream()
      } else {
        delegate.success(event as! T)
      }
    }
    queue.removeAll()
  }
}
